Include on-premise documentation as part of general Sigrid documentation (#392)
Software-Improvement-Group/sigridci: docs/organization-integration/onpremise-integration.md

- SIG provides the necessary images through a container registry. The section [obtaining Sigrid on-premise](#obtaining-sigrid-on-premise) contains more information on how you can obtain and update these Docker containers.
- Authentication is based on your identity provider, using [OpenID Connect](https://openid.net/developers/how-connect-works/). 
- Analyses are triggered from a build pipeline. The analysis results are then imported into a Postgres database, so they can be viewed in Sigrid.
- Large files are stored in an [S3-compatible object store](https://aws.amazon.com/s3/).

Some Sigrid on-premise features are *optional* based on your environment:

- The Open Source Health feature requires outbound internet access, to retrieve the latest version of all known vulnerabilities in open source libraries. 
- When viewing detailed analysis results, Sigrid displays relevant source code files within Sigrid. For this to work, a web-accessible code storage needs to be available. This integrates with Sigrid via [OAuth](https://oauth.net/2/). For this to work, the identity provider used for Sigrid authentication and for the code storage needs to be the same.

> ### Environment requirements

Reviewer: Do we want to specify versions? For example we have a minimum version requirement of kubernetes version 1.21  and Postgres version 13 but only aim to support non EOL versions of kubernetes (the last three releases)?

- Your infrastructure needs to support running applications on Kubernetes.
- You need to pull Sigrid container images from DockerHub.
  - If you do not allow outbound internet access, a custom air-gapped installation and update procedure needs to be designed, documented and followed.
- You need to support and manage a Postgres database service.
- You need to have a development platform with a continuous integration pipeline, since Sigrid will integrate with this pipeline. Supported development platforms are listed in the "Sigrid CI pipeline integration" in this documentation.
- You need to have an identity provider supporting OpenID Connect, since Sigrid will integrate with this for authentication.
- You need to allow outbound internet traffic in order to support Sigrid’s Open Source Health feature. Sigrid needs to connect to external sources to check for the latest vulnerability data.
  - If you do not allow outbound internet access, the Open Source Health feature is not available. The rest of Sigrid is unaffected.

> - For viewing source code within Sigrid, you need to provide a development platform that is integrated with th esame identity provider as Sigrid itself.

Reviewer: ```suggestion
- For viewing source code within Sigrid, you need to provide a development platform that is integrated with the same identity provider as Sigrid itself.
```

  - The view source functionality is optional, without this integration the rest of Sigrid is unaffected.

## Obtaining Sigrid on-premise

The Docker containers that form Sigrid on-premise are distributed via [DockerHub](https://hub.docker.com). You will receive an account that allows you to access the container registry. 

<img src="../images/onpremise-dockerhub.png" width="500" /> 

As explained above, Sigrid consists of several Docker containers. The container `sigrid-multi-analysis-import` will run directly from your development platform, all other containers are deployed to your Kubernetes cluster. These steps are explained in more detail in the following sections.

> ### Updating Sigrid on-premise to a new version

Reviewer: It is important that all components (APIs and analysis container) are update to the same version.

Author: Good one, will add.

- Analyses are triggered from a build pipeline. The analysis results are then imported into a Postgres database, so they can be viewed in Sigrid.
- Large files are stored in an [S3-compatible object store](https://aws.amazon.com/s3/).

Some Sigrid on-premise features are *optional* based on your environment:

- The Open Source Health feature requires outbound internet access, to retrieve the latest version of all known vulnerabilities in open source libraries. 
- When viewing detailed analysis results, Sigrid displays relevant source code files within Sigrid. For this to work, a web-accessible code storage needs to be available. This integrates with Sigrid via [OAuth](https://oauth.net/2/). For this to work, the identity provider used for Sigrid authentication and for the code storage needs to be the same.

### Environment requirements

- Your infrastructure needs to support running applications on Kubernetes.
- You need to pull Sigrid container images from DockerHub.
  - If you do not allow outbound internet access, a custom air-gapped installation and update procedure needs to be designed, documented and followed.
- You need to support and manage a Postgres database service.

> - You need to have a development platform with a continuous integration pipeline, since Sigrid will integrate with this pipeline. Supported development platforms are listed in the "Sigrid CI pipeline integration" in this documentation.

Reviewer 2: no longer relevant

- You need to have an identity provider supporting OpenID Connect, since Sigrid will integrate with this for authentication.

> - You need to allow outbound internet traffic in order to support Sigrid’s Open Source Health feature. Sigrid needs to connect to external sources to check for the latest vulnerability data.

Author: Interestingly, both originate from the original Word document that we used before moving this into the Sigrid documentation :) But you're right, let's consolidate it.

The on-premise Sigrid CI requires the `sigrid-multi-analysis-import` Docker container. The steps for including this container in your pipeline are different for each development platform. The next sections cover some common examples:

### GitHub pipeline integration

Your can add Sigrid CI on-premise to your GitHub pipeline by creating [GitHub Actions YAML files](https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions) inside of the `.github/workflows` directory.

```
name: sigrid-publish
on:
  schedule:
     - cron: "0 0 * * *"

jobs:
  sigridci:
    container:
      image: softwareimprovementgroup/sigrid-multi-analysis-import
      env:
        CUSTOMER: YOUR_SIGRID_CUSTOMER
        SYSTEM: YOUR_SIGRID_SYSTEM
        POSTGRES_HOST_AND_PORT: YOUR_DATABASE_HOST:5432
        POSTGRES_PASS: $SIGRID_REVIEW_ONPREM_POSTGRES_IMPORTER_PASS
    steps:

>       - ln -s $PWD /tmp/sources

Reviewer 2: So for github/gitlab/azure devops the symlink creation isn't necessary at all anymore, but we can also take this out later.

      - /usr/local/bin/all-the-things.sh
      - mv /tmp/analysis-results/sigridci .
```

Note this is a *generic* example, the specifics might be slightly different depending on how you have configured your on-premise GitHub.

This example will analyze your code on a daily basis. It is also possible to run the analysis after every change, and provide feedback in pull requests. Refer to the [Sigrid CI GitHub documentation](../sigridci-integration/github-actions.md) for more information.

### GitLab pipeline integration

Your can add Sigrid CI on-premise to your GitLab pipeline by editing the `.gitlab-ci.yml` configuration file:

```
sigridci-onpremise:
  stage: test
  image: softwareimprovementgroup/sigrid-multi-analysis-import
  entrypoint: [""]
  variables:
    CUSTOMER: YOUR_SIGRID_CUSTOMER
    SYSTEM: YOUR_SIGRID_SYSTEM
    POSTGRES_HOST_AND_PORT: YOUR_DATABASE_HOST:5432
    POSTGRES_PASS: $SIGRID_REVIEW_ONPREM_POSTGRES_IMPORTER_PASS
  script:

>     - ln -s $PWD /tmp/sources

Reviewer 2: Same here, maybe we can insert this in the all-the-things.sh script (without breaking if the symlink is already existing)

    - /usr/local/bin/all-the-things.sh
    - mv /tmp/analysis-results/sigridci .
  artifacts:
    reports:
      junit: sigridci/sigridci-junit-format-report.xml
    paths:
      - sigridci/index.html
      - sigridci/feedback.md
    expire_in: 1 week
    when: always
  rules:
    - if: $CI_PIPELINE_SOURCE == "schedule"
```

- This example assumes your pipeline includes a stage called `test`. If your pipeline is structured differently you will need to replace this with the name of the appropriate stage where you want Sigrid CI to run. 

> - Replace `YOUR_SIGRID_CUSTOMER` and `YOUR_SIGRID_SYSTEM` with your Sigrid customer account name and system name respectively. If you're not sure about your Sigrid customer account name, ask your Sigrid administrator. You can choose your own system name, though system names can only contain letters, numbers, and hyphens.

Reviewer 2: ```suggestion
- Replace `YOUR_SIGRID_CUSTOMER` and `YOUR_SIGRID_SYSTEM` with your Sigrid customer account name and system name, respectively. If you're not sure about your Sigrid customer account name, ask your Sigrid administrator. You can choose your own system name, though system names can only contain letters, numbers, and hyphens.
```

- Replace `YOUR_DATABASE_HOST` with your Postgres database host name. Your GitLab pipeline should be able to access this host.
- Use `$SIGRID_REVIEW_ONPREM_POSTGRES_IMPORTER_PASS` to configure the password of your Postgres database. Use [GitLab secrets](https://docs.gitlab.com/ee/ci/secrets/) to provide this password, *do not* hard-code this password in your YAML file.

Next, you will need to schedule *when* you're going to publish your system to Sigrid. Azure DevOps allows you to [schedule pipelines](https://learn.microsoft.com/en-us/azure/devops/pipelines/process/scheduled-triggers?view=azure-devops&tabs=yaml). You can use this to periodically publish your system to Sigrid, for example as a nightly build. Alternatively, you can use Sigrid CI on a more continuous basis by running the pipeline step after every commit.

## Functional differences in Sigrid on-premise

- The on-premise Sigrid distribution is single tenant. You cannot create your own "tenants", all systems and analyses will end up in your portfolio. That said, you can still use Sigrid’s user management to define which people should have access to which systems.

> - [Mendix Quallty and Security Management](https://www.softwareimprovementgroup.com/mendix-quality-and-security-management-powered-by-sigrid/) is only available if you are using Mendix in combination with GitHub or GitLab.

Reviewer 2: ```suggestion
- [Mendix Quality and Security Management](https://www.softwareimprovementgroup.com/mendix-quality-and-security-management-powered-by-sigrid/) is only available if you are using Mendix in combination with GitHub or GitLab.
```
